src/PlotlyEditor.js

```diff
@@ -35,23 +44,95 @@ class PlotlyEditor extends Component {
       graphDiv: gd,
       layout: gd.layout,
       locale: this.props.locale,
-      onUpdate: this.updateProp.bind(this),
+      onUpdate: this.handleUpdate.bind(this),
       plotSchema: this.plotSchema,
       plotly: this.props.plotly,
     };
   }
 
-  updateProp(event) {
+  handleUpdate({type, payload}) {
     const {graphDiv} = this.props;
-    if (this.props.onUpdate) {
-      this.props.onUpdate({graphDiv, ...event});
+
+    switch (type) {
+      case EDITOR_ACTIONS.UPDATE_TRACES:
+        if (this.props.onUpdateTraces) {
+          this.props.onUpdateTraces(payload);
+        }
+        for (let i = 0; i < payload.traceIndexes.length; i++) {
+          for (const attr in payload.update) {
+            const traceIndex = payload.traceIndexes[i];
+            const prop = nestedProperty(graphDiv.data[traceIndex], attr);
+            const value = payload.update[attr];
+            if (value !== void 0) {
+              prop.set(value);
+            }
+          }
+        }
+        if (this.props.onUpdate) {
+          this.props.onUpdate();
+        }
+        break;
+
+      case EDITOR_ACTIONS.UPDATE_LAYOUT:
+        if (this.props.onUpdateLayout) {
+          this.props.onUpdateLayout(payload);
+        }
+        for (const attr in payload.update) {
+          const prop = nestedProperty(graphDiv.layout, attr);
+          const value = payload.update[attr];
+          if (value !== void 0) {
+            prop.set(value);
+          }
+        }
+        if (this.props.onUpdate) {
+          this.props.onUpdate();
+        }
+        break;
+
+      case EDITOR_ACTIONS.ADD_TRACE:
+        if (this.props.onAddTrace) {
+          this.props.onAddTrace(payload);
+        }
+        graphDiv.data.push({x: [], y: []});
+        if (this.props.onUpdate) {
+          this.props.onUpdate();
+        }
+        break;
+
+      case EDITOR_ACTIONS.DELETE_TRACE:
+        if (this.props.onDeleteTrace) {
+          this.props.onDeleteTrace(payload);
+        }
+        if (payload.traceIndexes && payload.traceIndexes.length) {
+          graphDiv.data.splice(payload[0], 1);
+          if (this.props.onUpdate) {
+            this.props.onUpdate();
+          }
+        }
+        break;
+
+      case EDITOR_ACTIONS.DELETE_ANNOTATION:
+        if (this.props.onDeleteAnnotation) {
+          this.props.onDeleteAnnotation(payload);
+        }
+        if (isNumeric(payload.annotationIndex)) {
+          graphDiv.layout.annotations.splice(payload.annotationIndex, 1);
+          if (this.props.onUpdate) {
+            this.props.onUpdate();
+          }
+        }
+        break;
+
+      default:
+        throw new Error('must specify an action type to handleEditorUpdate');
     }
   }
 
   render() {
     return (
       <div className={bem()}>
         {this.props.graphDiv &&
+          this.props.graphDiv._fullLayout &&
```

Comment:
this is a big change, conceptually, because it allows us to pass a partially-constructed figure into the editor and the plot, then the plot finishes constructing it.